### PR TITLE
Add GlobalSubtyping pass

### PR DESCRIPTION
**Pull request by kripken**
This specializes the fields of structs based on the types written to them. That is,
if a field is of type A but in practice we always write some subtype B to it
then we can change the type of the field to that.

On j2wasm this manages to improve at least one field in 2% of types. Not a
large amount, but this does lead to further benefits in later opts (e.g. about a third
of the improvements are to turn a field non-nullable).

### CODE REVIEW

File: src/passes/GlobalSubtyping.cpp
Comment (ajklein) [inline src/passes/GlobalSubtyping.cpp:182]: Having stared at this for awhile I'm still having trouble understanding under what circumstances this happens (I think I'm not fully modeling the propagation steps above). Which test case(s) cover this?

**Review by tlively — COMMENTED** [review]
Posting all my pending comments from when I started taking a look last week, but I haven't finished reading through this yet.

File: src/passes/GlobalSubtyping.cpp
Comment (kripken) [reply src/passes/GlobalSubtyping.cpp:182]: The testcase beginning on line 422 covers this,

https://github.com/WebAssembly/binaryen/pull/4306/files#diff-f4453daffbce437bfcb3bf0a2117bfdc32fe747707ecfc546f3771c0d34b38e3R422-R466

This is a pretty subtle point and the code should explain it better. I'll push a comment with an example in a minute.

Comment (kripken) [reply src/passes/GlobalSubtyping.cpp:182]: (see my response to @ajklein right below, and the comment I just pushed)

File: src/passes/GlobalSubtyping.cpp
Comment (kripken) [reply src/passes/GlobalSubtyping.cpp:31]: Ah, sorry about this namespace stuff, I think it's a leftover from some debugging that I missed...

File: src/passes/GlobalSubtyping.cpp
Comment (kripken) [reply src/passes/GlobalSubtyping.cpp:74]: Sounds good. That's a refactoring that crosses a few other existing files, so I'll do it separately?

File: src/passes/GlobalSubtyping.cpp
Comment (kripken) [reply src/passes/GlobalSubtyping.cpp:171]: Note that while the field has no writes, but a super's field might - so we have no guarantee of GTO being able to remove it.

Overall, we have to pick *some* type for the field, which is all this does - we have no better idea of a type than the old  type, which we just leave there.

I'll add more comments to help here.

**Comment by kripken** [comment]
Thanks for the feedback, I pushed some more comments and clarifications now. I also fixed the merge conflict.

File: src/passes/GlobalSubtyping.cpp
Comment (aheejin) [inline src/passes/GlobalSubtyping.cpp:114]: Basic question: Why does this pass (and `GlobalTypeOptimization` too) require nominal typing?

**Review by aheejin — COMMENTED** [review]
I started to read this but I'm not sure if I'll finish until you merge, and at the moment I only have a few simple nits and questions. I'll post them anyway but don't wait for me for the merge.

File: src/passes/GlobalSubtyping.cpp
Comment (kripken) [reply src/passes/GlobalSubtyping.cpp:114]: Nominal typing lets us see exactly which types are subtypes of others. Technically we could compute that for structural subtyping too, but it takes a lot more work (we'd need to duplicate the logic in the canonicalization code, I assume), and also it would be less effective anyhow as when it merges types it limits optimization opportunities (since we can only optimize the merged type in one way, and not each of the original types differently as is best for it).

**Comment by kripken** [comment]
Prelude PR: https://github.com/WebAssembly/binaryen/pull/4317

**Review by ajklein — COMMENTED** [review]
Looks good to me, but I naturally defer to @tlively and @aheejin on proper approval here.

File: test/lit/passes/global-subtyping.wast
Comment (tlively) [inline test/lit/passes/global-subtyping.wast:225]: What is `GP`?

**Review by tlively — APPROVED** [review]
Very cool to see this working.

File: src/passes/GlobalSubtyping.cpp
Comment (kripken) [reply src/passes/GlobalSubtyping.cpp:208]: Interesting, yeah, maybe this should really be on fields. I'll add a TODO.

File: test/lit/passes/global-subtyping.wast
Comment (kripken) [reply test/lit/passes/global-subtyping.wast:225]: Clearly I spend too much time online... "GrandParent" (or really "Grandparent Post[er]"). That is, not the preceding testcase but one before it. I'll clarify.

File: test/lit/passes/global-subtyping.wast
Comment (kripken) [reply test/lit/passes/global-subtyping.wast]: Sounds good, added a mention of $child-B. That shows that it has been updated too, as expected.

File: test/lit/passes/global-subtyping.wast
Comment (tlively) [reply test/lit/passes/global-subtyping.wast]: I've only ever seen "OP" in the wild. Clearly I need to spend more time online!

**Pull request merged**
